File: job-matcher/src/config/config-loader.js
const fs = require('fs');
const path = require('path');
const yaml = require('js-yaml');

class ConfigLoader {
  static load(configPath = './config/jobs.yaml') {
    try {
      const fullPath = path.resolve(configPath);
      
      if (!fs.existsSync(fullPath)) {
        throw new Error(`Configuration file not found: ${fullPath}`);
      }
      
      const fileContent = fs.readFileSync(fullPath, 'utf8');
      let config = yaml.load(fileContent);
      
      // Process environment variables in config
      config = this.processEnvVariables(config);
      
      // Validate required fields
      this.validateConfig(config);
      
      // Resolve relative paths
      config.resume_path = path.resolve(config.resume_path);
      config.output_path = path.resolve(config.output_path);
      
      return config;
    } catch (error) {
      console.error('Error loading configuration:', error.message);
      process.exit(1);
    }
  }
  
  static processEnvVariables(obj) {
    if (typeof obj === 'string') {
      return obj.replace(/\$\{([^}]+)\}/g, (match, envVar) => {
        const value = process.env[envVar];
        if (value === undefined) {
          console.warn(`Warning: Environment variable ${envVar} is not set`);
          return match;
        }
        return value;
      });
    }
    
    if (Array.isArray(obj)) {
      return obj.map(item => this.processEnvVariables(item));
    }
    
    if (typeof obj === 'object' && obj !== null) {
      const result = {};
      for (const [key, value] of Object.entries(obj)) {
        result[key] = this.processEnvVariables(value);
      }
      return result;
    }
    
    return obj;
  }
  
  static validateConfig(config) {
    const required = ['resume_path', 'output_path', 'llm', 'job_boards'];
    
    for (const field of required) {
      if (!config[field]) {
        throw new Error(`Missing required configuration field: ${field}`);
      }
    }
    
    if (!Array.isArray(config.job_boards) || config.job_boards.length === 0) {
      throw new Error('Configuration must include at least one job_board');
    }
    
    if (!config.llm.provider) {
      throw new Error('LLM provider must be specified');
    }
    
    const validProviders = ['groq', 'ollama', 'openrouter', 'gemini'];
    if (!validProviders.includes(config.llm.provider)) {
      throw new Error(`Invalid LLM provider: ${config.llm.provider}`);
    }
    
    for (const board of config.job_boards) {
      if (!board.name || !board.type || !board.url) {
        throw new Error('Each job board must have name, type, and url');
      }
    }
  }
}

module.exports = ConfigLoader;

File: job-matcher/src/config/config_loader.py
import os
import re
import yaml
from pathlib import Path


class ConfigLoader:
    REQUIRED_FIELDS = ["resume_path", "output_path", "job_boards"]

    @classmethod
    def load(cls, config_path=None):
        if config_path is None:
            config_path = Path(__file__).parent.parent.parent / "config" / "jobs.yaml"
        config_path = Path(config_path).resolve()

        if not config_path.exists():
            raise FileNotFoundError(f"Configuration file not found: {config_path}")

        with open(config_path) as f:
            config = yaml.safe_load(f)

        config = cls._process_env_vars(config)
        config = cls._apply_env_overrides(config)
        cls._validate(config)

        config["resume_path"] = str(Path(config["resume_path"]).resolve())
        config["output_path"] = str(Path(config["output_path"]).resolve())
        config["ats_cache_db_path"] = str(
            Path(config.get("ats_cache_db_path", "./output/ats-cache.sqlite")).resolve()
        )

        return config

    @classmethod
    def _process_env_vars(cls, obj):
        if isinstance(obj, str):
            def replacer(m):
                env_var = m.group(1)
                value = os.environ.get(env_var)
                if value is None:
                    print(f"Warning: Environment variable {env_var} is not set")
                    return m.group(0)
                return value
            return re.sub(r"\$\{([^}]+)\}", replacer, obj)
        if isinstance(obj, list):
            return [cls._process_env_vars(item) for item in obj]
        if isinstance(obj, dict):
            return {k: cls._process_env_vars(v) for k, v in obj.items()}
        return obj

    @classmethod
    def _validate(cls, config):
        for field in cls.REQUIRED_FIELDS:
            if not config.get(field):
                raise ValueError(f"Missing required configuration field: {field}")

        boards = config.get("job_boards", [])
        if not isinstance(boards, list) or len(boards) == 0:
            raise ValueError("Configuration must include at least one job_board")

        llm = config.get("llm", {})
        if llm.get("provider") and llm.get("provider") != "openrouter":
            raise ValueError(f"Unsupported llm.provider: {llm['provider']}. Supported: openrouter")
        if llm.get("provider") == "openrouter":
            api_key = llm.get("api_key") or os.environ.get("OPENROUTER_KEY")
            if not api_key:
                raise ValueError("OpenRouter API key required. Set llm.api_key or OPENROUTER_KEY env var.")

        for board in boards:
            if not board.get("name") or not board.get("type") or not board.get("url"):
                raise ValueError("Each job board must have name, type, and url")

    @classmethod
    def _apply_env_overrides(cls, config):
        result = dict(config)
        result["scoring"] = dict(config.get("scoring", {}))

        force = os.environ.get("JOB_MATCHER_FORCE_RECOMPUTE", "").strip().lower()
        if force:
            result["scoring"]["force_recompute"] = force in ("1", "true", "yes", "on")

        if os.environ.get("JOB_MATCHER_OUTPUT_PATH"):
            result["output_path"] = os.environ["JOB_MATCHER_OUTPUT_PATH"]

        return result
